torchtune/models/clip/_tokenizer.py

```diff
@@ -0,0 +1,211 @@
+# Copyright (c) Meta Platforms, Inc. and affiliates.
+# All rights reserved.
+#
+# This source code is licensed under the BSD-style license found in the
+# LICENSE file in the root directory of this source tree.
+import gzip
+import html
+from pathlib import Path
+from typing import List
+
+import ftfy
+import regex as re
+import torch
+
+from torchtune.modules.tokenizers._utils import BaseTokenizer
+
+
+class CLIPTokenizer(BaseTokenizer):
+    """
+    Text tokenizer for CLIP.
+
+    Based on the official implementation here:
+    https://github.com/openai/CLIP/blob/main/clip/simple_tokenizer.py
+
+    Args:
+        vocab_path (Path): the path to the CLIP vocab file
+        max_seq_len (int): the context length (all CLIP models use 77)
+        truncate (bool): whether to truncate the text when longer than max_seq_len
+    """
+
+    def __init__(self, vocab_path: Path, max_seq_len: int = 77, truncate: bool = True):
+        self.max_seq_len = max_seq_len
+        self.truncate = truncate
+
+        self.byte_encoder = _bytes_to_unicode()
+        self.byte_decoder = {v: k for k, v in self.byte_encoder.items()}
+
+        with gzip.open(vocab_path) as f:
+            merges = f.read().decode("utf-8").split("\n")
+        merges = [tuple(merge.split()) for merge in merges[1:48895]]
+
+        vocab = list(self.byte_encoder.values())
+        vocab.extend([v + "</w>" for v in vocab])
+        vocab.extend(["".join(merge) for merge in merges])
+        vocab.extend(["<|startoftext|>", "<|endoftext|>"])
+
+        self.encoder = {word: i for i, word in enumerate(vocab)}
+        self.decoder = {v: k for k, v in self.encoder.items()}
+        self.bpe_ranks = {merge: i for i, merge in enumerate(merges)}
+
+        self.pat = re.compile(
+            r"""<\|startoftext\|>|<\|endoftext\|>|'s|'t|'re|'ve|'m|'ll|'d|[\p{L}]+|[\p{N}]|[^\s\p{L}\p{N}]+""",
+            re.IGNORECASE,
+        )
+
+        self.sot_token = self.encoder["<|startoftext|>"]
+        self.eot_token = self.encoder["<|endoftext|>"]
+        self.pad_token = self.eot_token
+
+        self.cache = {
+            "<|startoftext|>": "<|startoftext|>",
+            "<|endoftext|>": "<|endoftext|>",
+        }
+
+    def __call__(self, texts: List[str]) -> torch.Tensor:
```

Comment:
I added an extra method that's not in the tokenizer interface to make it easy to get a batch of token sequences for CLIP:
`tokens_tensor = tokenizer(list_of_texts)`.

Should this function be moved out of the tokenizer class or renamed or anything?

Reply:
why would the clip tokenizer expect message objects?

Reply:
We discussed this offline. Our decoder tokenizers expect messages, which is what is returned in the recipes. Since this is for an encoder though, it is fine for now to expect text directly since presumably the encoder tokenizer would not be directly exposed in the recipe.